### DISCUSSION TITLE
Dark Mode

### DISCUSSION THREAD

**Discussion by thatmattlove** [netbox-community/netbox#5782] [Ideas]
Hello,

I've taken a look at #3599, and it looks like dark mode would be a welcome feature. Would the maintainers have any opposition to me running with a migration to Bootstrap 5 (and related dependencies), as well as an implementation of a dark theme based purely on CSS media queries? That way, Netbox's theme would simply align with the user's defined color preference.

I've got a fair amount of experience with front-end languages/concepts, you can check [hyperglass](https://github.com/checktheroads/hyperglass), [my company's website](https://github.com/stellaraf/site), or the [website](https://github.com/48ix/site) of the IX I run if you'd like.

If this is something that would be welcome by the community, I'd be willing to take it on.

**Comment by jsenecal** [comment]
I've seen your work, awesome job! I would definitely appreciate a UI refresh and dark versions of Netbox but it seems like there's not a lot of interest unfortunately as the aforementioned issue got very little attention. (Thanks for hyperglass though)

**Reply by thatmattlove** [reply]
I would tend to agree, given this thread sat dormant for 2 weeks :)

Perhaps the approach is to focus on upgrading Bootstrap - Netbox is still using 3.x. The addition of CSS variable support in 4 & 5 would make the addition of dark mode trivial. Do you know if this is being discussed? I'd be more than happy to get started, if no one else has.

**Reply by xkilian** [reply]
I would suggest raising the issue and propose to do the work on upgrading bootstrap.

**Reply by jeremystretch** [reply]
We need to move from Bootstrap 3 to Bootstrap 5. (Yes, it's that old.)

**Reply by thatmattlove** [reply]
From a maintainer perspective, are there any hard and fast browser requirements that Netbox must support? Namely, IE 11? This would be my only "concern" with moving to BS5 - IE support is dropped. This is common with any framework these days, but I wasn't sure if there were any users (or support-paying customers) that had a requirement to support IE 11. If not, I'll open an issue and volunteer to do the work as @xkilian suggested.

**Reply by jeremystretch** [reply]
That's a good question. I'm not up to speed enough on where browsers are these days regarding different features. We're probably okay with whatever BS5 supports given its popularity, but I'm open to feedback.

**Reply by jeremystretch** [reply]
I've opened #5893 to give this more visibility. I don't know specifically what BS5 support for dark mode looks like, but I have to imagine it's better than whatever options we have currently.

**Reply by thatmattlove** [reply]
Sweet - I responded there as well, but I'd be happy to help with this in whatever way would be most beneficial to the community. I don't have a ridiculous amount of free time to dedicate, but I'll put in as much as I can.

Re: dark mode - easily accomplished with the addition of CSS variable support (or compiling Sass, or both).

**Comment by DanSheps** [comment]
> I've got a fair amount of experience with front-end languages/concepts, you can check hyperglass, my company's website, or the website of the IX I run if you'd like.

Would you be interested in doing a API driven frontend for us?  lol

**Reply by thatmattlove** [reply]
If that's the direction the maintainers want to go, instead of templates and traditional HTML/CSS, I would _love_ to help with it, for sure!

**Reply by candlerb** [reply]
And the logical conclusion would be to put all the back-end logic in Postgres and expose it using [PostgREST](https://postgrest.org/)...

**Reply by thatmattlove** [reply]
The most common thing to do is use DRF. Coupled with React, and data fetching/mutation libraries like react-query or swr, it's a very cool stack.

**Reply by jeremystretch** [reply]
I'm all for an API-driven UI; we've long envisioned this as effecting "NetBox v3.0." But I worry about trying to take on too much too fast. @checktheroads do you think it makes sense to tackle the UI refresh and conversion to BS5 in v2.11, with an eye toward supporting a future API-backed frontend, and then start on that migration in the following release? Or is it more practical to take on all at once?

**Reply by thatmattlove** [reply]
That's a tough one. I can say that going the SPA frontend API backend route for NetBox would be a pretty considerable undertaking - albeit worth it, in the end.

There's a lot to consider - authentication would completely change, DRF serializers modified/added that potentially don't exist today (I'm admittedly not overly familiar with the NetBox code structure, yet), new OS dependencies/installation processes (i.e. NodeJS), bundle/build processes that need to be built into startup, to say nothing of the development process for maintainers being _very_ different. But again, worth it.

After thinking that through a bit, that definitely feels like a more "3.0 major breaking change" type thing that something to slip into a minor release. My opinion would be to stay on course with BS5 and migrating away from jQuery - this alone will help a future migration to SPA/API.

**Reply by candlerb** [reply]
> DRF serializers modified/added that potentially don't exist today

That's the big one.  There's quite a lot of functionality in the UI which can't be replicated via the API, or at least, not efficiently.  In particular, there are various search filters that aren't exposed via the API, and little or no control over what related objects are fetched.

This could be implemented by a whole bunch of ad-hoc REST API endpoints customised to the needs of the UI (ugh).  Or it could be done generically with something like GraphQL.  Or it could be done by defining RESTful ways to do certain custom queries, as PostgREST does - e.g. see [filtering](https://postgrest.org/en/v7.0.0/api.html#horizontal-filtering-rows), [resource embedding](https://postgrest.org/en/v7.0.0/api.html#resource-embedding).  Personaly, I don't really see the value in putting a layer over SQL: the Salesforce "REST" API also exposes its own [SQL-based query language](https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql.htm).

In any case, I think we agree that the API is going to change significantly, which seems like "3.0" to me too.

**Reply by jeremystretch** [reply]
> Personaly, I don't really see the value in putting a layer over SQL

There's a ton of logic and validation within the REST API. Enforcement of object-based permissions alone necessitates this.

**Reply by thatmattlove** [reply]
> Enforcement of object-based permissions alone necessitates this.

Agreed - if the app had zero requirements for authentication, I could see the value in removing the extra layer. But DRF and one of the DRF JWT libraries (can't recall the name at the moment) make it very easy to flexibly handle authentication/authorization for each view.

I _do_ however like the idea of using GraphQL - it would make the querying and filtering much less obtuse in those cases where fine granularity is needed.

> In particular, there are various search filters that aren't exposed via the API, and little or no control over what related objects are fetched.

Indeed, I anticipate needing to add a lot of UI-specific serialzers. Another reason I think putting this endeavor on the 3.0 roadmap is probably wise :)
